Microsoft.Toolkit/Deferred/DeferredEventArgs.cs

```diff
@@ -0,0 +1,91 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.ComponentModel;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.Toolkit.Deferred
+{
+    /// <summary>
+    /// <see cref="EventArgs"/> which can retrieve a <see cref="EventDeferral"/> in order to process data asynchronously before an <see cref="EventHandler"/> completes and returns to the calling control.
+    /// </summary>
+    public class DeferredEventArgs : EventArgs, IDisposable
+    {
+        /// <summary>
+        /// Gets a new <see cref="DeferredEventArgs"/> to use in cases where no <see cref="EventArgs"/> wish to be provided.
+        /// </summary>
+        public static new DeferredEventArgs Empty => new DeferredEventArgs();
+
+        private readonly object _eventDeferralLock = new object();
+
+        private EventDeferral? _eventDeferral;
+        private bool _disposed;
+
+        /// <summary>
+        /// Finalizes an instance of the <see cref="DeferredEventArgs"/> class.
+        /// </summary>
+        ~DeferredEventArgs() => Dispose(false);
+
+        /// <summary>
+        /// Returns an <see cref="EventDeferral"/> which can be completed when deferred event is ready to continue.
+        /// </summary>
+        /// <returns><see cref="EventDeferral"/> instance.</returns>
+        public EventDeferral GetDeferral()
+        {
+            lock (_eventDeferralLock)
+            {
+                return _eventDeferral ?? (_eventDeferral = new EventDeferral());
+            }
```

Comment:
I'm a bit confused as to how this would work when more than one subscriber to the event asks for a deferral, wouldn't the object that raised the event just wait until the first one that completes the (same) deferral signals a completion?
As in, imagine this scenario:

- An object `sender` raises an event with somr `DeferredEventArgs` argument.
- A subscriber `sub1` is the first in the invocation list, so starts executing its handler.
- `sub1` calls `GetDeferral`, then starts some asynchronous operations, yielding back.
- The multicast delegate keeps going and gets to the second target in the invocation list.
- `sub2` calls `GetDeferral`, then starts some asynchronous operation and yields back.
...
- `sub2`'s operation completes first. It gets control back and completes the deferral.
- `sender` keeps going as the deferral is complete.

...Meanwhile `sub1` is still awaiting, but nobody is waiting for it 😢

If this is by design and we only want to support at most one receiver with a deferral, I think we should document so, and possibly detect cases where this happens and possibly throw, or indicate the intended behavior somehow 🤔

Reply:
Good call out @Sergio0694, we should probably file a new issue for the future. This was the same in the [original library](https://github.com/pedrolamas/DeferredEvents/blob/master/src/DeferredEvents/DeferredEventArgs.cs#L18) as well.

Though, I think in most cases we've seen at a UI layer, there's usually only one subscriber, but there's no guarantee of that for sure based on C# conventions.

That could be something we clean-up as an internal implementation detail in the future, eh? I don't think we need to modify anything later? I think the main thing is if the base usage of the event raiser would know what to do with multiple subscribers, as here we just expect the event args to be modified:

https://github.com/windows-toolkit/WindowsCommunityToolkit/blob/189e91aa13437a276dda6667894cd0c084e6b27b/Microsoft.Toolkit.Uwp.UI.Controls.Input/TokenizingTextBox/TokenizingTextBox.cs#L436-L447

Reply:
Please bear with me as I've written that code quite a few years ago, but I believe at the time I made sure that you don't call the sub2 before sub1 is complete!

I'll check this better tomorrow, it's now 0h here in the UK...

Reply:
I've double-checked and the code is correct: after each handler invocation, we will then call `eventArgs.GetCurrentDeferralAndReset();` so that we get the current `TaskCompletionSource` instance from that invocation and reset it before we move on to the next handler - so each handler has his own chance to call for deferral.

So between each handler's call to `GetDeferral()` and the next handler, we call the `GetCurrentDeferralAndReset()` that will ensure that each one has a unique awaitable instance.

Now what this means is that we can have multiple handlers executing at the same time, but we will still wait for all of them to finish before moving ahead (that's where that `Task.WhenAll` comes along)

Reply:
Ah, sorry I assumed you would just invoke the handler normally and simply await on the args afterwards. I see from the code that @michael-hawker linked that you're using a custom extension to sequentially run the registered handlers instead, yeah that'll work fine for now. That approach does have the limitation that handlers will not be able to run sequentially though, so eg. if one takes a long time it'll also delay all the others, but this is consistent with the previous behavior anyway so we might as well address this in another PR in the future if it becomes a problem or just leave it as is, which is fine for more scenarios anyway 🙂

Reply:
Indeed you are correct, it will run the even handlers in parallel and not wait for the completion of each before running the next!

However, if we do want to do that, it is quite trivial to add an optional boolean parameter to the `InvokeAsync` method with an alternative execution (so if that parameter is set to true, we would ensure await immediately instead of using the `Task.WhenAll` in the end!